src/file_system/mod.rs

```diff
@@ -0,0 +1,345 @@
+        Repo::new()
+    }
+
+    pub fn list_directory(&self) -> Vec<(Label, SystemType)>
+    where
+        Repo: Clone,
+    {
+        self.entries
+            .iter()
+            .cloned()
+            .filter_map(|entry| match entry {
+                DirectoryContents::SubDirectory(dir) => Some(SystemType::is_directory(dir.label)),
+                DirectoryContents::File(file) => Some(SystemType::is_file(file.filename)),
+                DirectoryContents::Repo(_) => None,
+            })
+            .collect()
+    }
+
+    pub fn find_file(&self, path: Path) -> Option<File>
+    where
+        Repo: Clone + std::fmt::Debug,
+    {
+        let (path, filename) = path.split_last();
+        let path = NonEmpty::from_slice(&path);
+
+        let search_directory = match path {
+            None => Some(self.clone()),
+            Some(p) => self.find_directory(Path(p)),
+        };
+
+        search_directory.and_then(|dir| dir.file_in_directory(&filename))
+    }
+
+    pub fn find_directory(&self, path: Path) -> Option<Self>
+    where
+        Repo: Clone,
+    {
+        let mut search_directory = Some(self.clone());
+        let (label, labels) = path.split_first();
+        if *label == self.label {
+            for label in labels {
+                match search_directory {
+                    None => return None,
+                    Some(dir) => {
+                        search_directory = dir.get_sub_directory(&label);
+                    }
+                }
+            }
+            search_directory
+        } else {
+            None
+        }
+    }
+
+    // TODO(fintan): This is going to be a bit trickier so going to leave it out for now
+    #[allow(dead_code)]
+    pub fn fuzzy_find(_label: Label) -> Vec<Self> {
+        unimplemented!()
+    }
+
+    fn get_sub_directories(&self) -> Vec<Self>
+    where
+        Repo: Clone,
+    {
+        self.entries
+            .iter()
+            .filter_map(|entry| match entry {
+                DirectoryContents::SubDirectory(dir) => Some(*dir.clone()),
+                DirectoryContents::File(_) => None,
+                DirectoryContents::Repo(_) => None,
+            })
+            .collect()
+    }
+
+    fn get_sub_directory(&self, label: &Label) -> Option<Self>
+    where
+        Repo: Clone,
+    {
+        self.get_sub_directories()
+            .iter()
+            .cloned()
+            .find(|directory| directory.label == *label)
+    }
+
+    fn file_in_directory(&self, label: &Label) -> Option<File> {
+        for entry in self.entries.iter() {
+            match entry {
+                DirectoryContents::File(file) => {
+                    if file.filename == *label {
+                        return Some(file.clone());
+                    } else {
+                        continue;
+                    }
+                }
+                DirectoryContents::SubDirectory(_) => continue,
+                DirectoryContents::Repo(_) => continue,
```

Comment:
You probably know this, but you can just do `DirectoryContents::Repo(_) => {},`

Reply:
I did not know this. I wonder what's clearer though :thinking:

Reply:
well, to me, since `continue` is the default, you don't need to spell it out, unless there's code after the `continue` that you want to skip over. But here it would continue regardless.